### ISSUE TITLE
Super Mario Eclipse audio streaming issue

### ISSUE THREAD

**Issue by SpyrooftheNight** [FIX94/Nintendont#1228]
Super Mario Eclipse has a lot of its new music go silent in Nintendont on Wii and Wii U. Something to do with audio streaming because turning that off on Swiss on GameCube causes the exact same issue.

**Comment by staringstar** [comment]
Someone already notified the Nintendon't devs here https://github.com/FIX94/Nintendont/issues/1182 about the issue, but it seems like it went unnoticed.

While Im no expert on Nintendont or GC/Wii in general, fixing this issue seems like it would be a sizeable effort so Im not getting my hopes up for a fix coming soon.

**Comment by SpyrooftheNight** [comment]
I won't get my hopes up either but my only way to play on my CRT is GameCube or Wii and I was experiencing slowdowns on GameCube not on Wii so if they could get it working it would be the best way to play the game authentically.

**Comment by carnage702** [comment]
the thing here is nintendont is not an emulator, its more like a bridge  between gc calls and wii hardware and redirects and changes those calls  so that the wii/wiiu play them using their hardware.

GC Homebrew does not work on nintendont because of that becuase it doesnt use the expect functions to do so so they dont get redirected correctly same for some mods.

nintendont patches patterns and functions from original nintendo sdk and homebrew just doesn't follow any rules, for nintendont to use audio streaming all the official functions and calls would need to be followed like they are on the other audisotreaming games which they arent so nintendont cant patch this mod there really isnt a fix that could be done unless you would manually patch every address for every song and hard coded it on nintendont and forced it in instead of naturally playing it like the other audiostreaming games do and i dont think anyone would do that complex work tbh.

**Comment by staringstar** [comment]
I just checked SMSBetterEngine's audio system code and it seems like it does import system functions to do its job, so a way to fix this can probably be found.

Something else that seems notable is that that engine's code does seem to not be inserted into the game's executable (or at least not until runtime), so no wonder that Nintendont can't detect and patch the underlying patterns.

**Comment by carnage702** [comment]
> I just checked SMSBetterEngine's audio system code and it seems like it does import system functions to do its job, so a way to fix this can probably be found.
> 
> Something else that seems notable is that that engine's code does seem to not be inserted into the game's executable (or at least not until runtime), so no wonder that Nintendont can't detect and patch the underlying patterns. The SMSBetterEngine module is inserted into the game by a mod loader called Kuribo ( altough Im not entirely sure if this is correct ) so a patch could be done by modifying SMSBetterEngine or Kuribo to do their own patches for Nintendont compatibily or something akin to Majora's Mask Nintendont patches.

Like I said it's a complex codding change so I don't know if anyone is willing to put the effort to learn nintendont code to even know what to fix.

**Comment by SpyrooftheNight** [comment]
So does that mean that the Eclipse team would have to fix their files?

**Comment by Extrems** [comment]
> the thing here is nintendont is not an emulator, its more like a bridge between gc calls and wii hardware and redirects and changes those calls so that the wii/wiiu play them using their hardware.

Maybe you should make it an emulator then since that worked for Swiss.

**Comment by carnage702** [comment]
> > the thing here is nintendont is not an emulator, its more like a bridge between gc calls and wii hardware and redirects and changes those calls so that the wii/wiiu play them using their hardware.
> 
> Maybe you should make it an emulator then since that worked for Swiss.

Swiss runs GC homebrew nintendont does not, it was the way nintendont was made and how it works if a function doesn't do what is expected on the exact place then it won't work

**Comment by madsl** [comment]
I really hope Nintendont will add this feature, because there's no way of getting swiss-gc to run on a Wii U. I even tried booting swiss from Nintendont just to make sure...

**Comment by JoshuaMKW** [comment]
> Someone already notified the Nintendon't devs here #1182 about the issue, but it seems like it went unnoticed.
> 
> While Im no expert on Nintendont or GC/Wii in general, fixing this issue seems like it would be a sizeable effort so Im not getting my hopes up for a fix coming soon.

If you can point me to _exactly_ what types of patches I need to do in the code, I'll consider it.

**Comment by VGBCP3rs0na** [comment]
I can't even hear the music in Erto Rock!!! This is bull!

**Comment by ReiulFox** [comment]
Please when a fix for this?

**Comment by carnage702** [comment]
> Please when a fix for this?

when you find someone to code it, there are no active nintendont devs in like 3 years that would know about audiostreaming.

**Comment by Extrems** [comment]
I made an accidental discovery that may offer an alternative explanation as to why this isn't working.
https://github.com/emukidid/swiss-gc/issues/963

Somebody mixed up `OSGetTime` and `__OSGetSystemTime`.

**Comment by ReiulFox** [comment]
With the discovery of Extrems could this error be solved?

**Comment by Extrems** [comment]
It could be solved by the Eclipse Team within their own mod, or worked around by Nintendont. The bug is in the mod.

**Comment by JoshuaMKW** [comment]
> It could be solved by the Eclipse Team within their own mod, or worked around by Nintendont. The bug is in the mod.

I feel like this take is disingenuous, for a couple of reasons. Firstly, because it works with Swiss due to a more robust approach. Secondly, because the only reason why it fails on Nintendont is because it applies its patches before the Super Mario Eclipse code is dynamically loaded and linked to the game in the same manner as a DLL. The problem lies in Nintendont's assumption that DVD streaming code would never be loaded in such a way.

**Comment by eku** [comment]
> The problem lies in Nintendont's assumption that DVD streaming code would never be loaded in such a way.

If you know what the cause is, why not provide a PR to fix it?

**Comment by Extrems** [comment]
> Firstly, because it works with Swiss due to a more robust approach.

Please explain why your code break when monotonic system time isn't reasonably close to wallclock time.

**Comment by ajbeas2015** [comment]
Is there a way to download and setup Swiss for the Wii to play Mario Eclipse with custom audio?

**Comment by JoshuaMKW** [comment]
> > Firstly, because it works with Swiss due to a more robust approach.
> 
> Please explain why your code break when monotonic system time isn't reasonably close to wallclock time.

I can't give you an answer to this because my DVD code doesn't use system time at all. Also all of my DVD streaming code was reverse engineered from Finding Nemo GC and additionally cross referenced with the YAGCD website.

**Comment by JoshuaMKW** [comment]
> > The problem lies in Nintendont's assumption that DVD streaming code would never be loaded in such a way.
> 
> If you know what the cause is, why not provide a PR to fix it?

I have as of now. Nintendont was the problem.

**Comment by JoshuaMKW** [comment]
> > > It's sad that this hasn't been resolved or a solution sought....
> > 
> > I solved it now, it's just a matter of the fixes being merged and a new release put out.
> 
> Alright. Tell me what to do and the specifics. Merging the fixes huh? Tell the people on GameBanana that and also Spike and the others on Retro Game Talk to put the new rom there. Finally a solution.

I'm a little confused, do you understand that I am "the people on GameBanana" that you are referring to? I also don't condone the distribution of pre patched roms. If I recommended it, Nintendo could take us down for it.

I'm glad you are excited for this problem to be resolved but please be patient and slow your roll a bit. We are doing the best we can to make sure this is done properly.

**Comment by Drflash55** [comment]
@carnage702 With the merge of #1322, may this issue also be closed as well? Once the new build of Nintendont is widely made available publicly, this won't have to stay up any longer.

**Comment by VGBCP3rs0na** [comment]
I mean the people who made Mario Eclipse. Tell them and also those in the Eclipse Discord too. If this doesn't fix the audio, I don't know what will.

**Comment by VGBCP3rs0na** [comment]
> [@carnage702](https://github.com/carnage702) With the merge of [#1322](https://github.com/FIX94/Nintendont/pull/1322), may this issue also be closed as well? Once the new build of Nintendont is widely made available publicly, this won't have to stay up any longer.

I don't think we should until the issue is fixed unless a new issue arises or this isn't it. I'm just as eager to fix this issue as you all are.

**Comment by madsl** [comment]
What you mean? Nintendont works now with the latest build and Eclipse 1.0.6, I hear the custom music just fine.

**Comment by ReiulFox** [comment]
> What you mean? Nintendont works now with the latest build and Eclipse 1.0.6, I hear the custom music just fine.

Where i can download the new build of nintendont where i can hear the custom music in eclipse?

**Comment by madsl** [comment]
@carnage702 commited a new loader.dol file yesterday, so you download it as you would normally do from this repo

**Comment by ReiulFox** [comment]
> [@carnage702](https://github.com/carnage702) commited a new loader.dol file yesterday, so you download it as you would normally do from this repo

I download it and i am playing right now, but i found a problem, The custom music stops playing after a while

**Comment by ReiulFox** [comment]
I don't know if it's the version; I'm playing 1.0.4 because the current version just shows a black screen. Someone said it was because a real memory card wasn't detected.

**Comment by JoshuaMKW** [comment]
Y'all are gonna drive me crazy. I already said multiple times at multiple places that y'all need to wait until Eclipse v1.1.0 comes out. Stop being shocked things aren't running smoothly when you are circumventing my advice to shortcut your way about.

**Comment by ReiulFox** [comment]
> Y'all are gonna drive me crazy. I already said multiple times at multiple places that y'all need to wait until Eclipse v1.1.0 comes out. Stop being shocked things aren't running smoothly when you are circumventing my advice to shortcut your way about.

Sorry for the inconvenience, I understand, I'll be patient until Eclipse v1.1.0 is released. It's incredible that you were able to fix this bug, I apologize again. It's logical that there will be bugs since you'll add the optimal fixes in the new version. By the way, I love your mod, it's very replayable and you can do a lot of things

**Comment by pokemonfan6498** [comment]
even though nintendon't released new version this year being 6.504 it still has same issue despite having listing for fixed audio streaming

**Comment by carnage702** [comment]
> even though nintendon't released new version this year being 6.504 it still has same issue despite having listing for fixed audio streaming

the mario eclipse dev team said only with the new version of eclipse will it work with nintendont, this was a prep for that upcoming version, dont know if they released the new version already or not.

**Comment by pokemonfan6498** [comment]
> > even though nintendon't released new version this year being 6.504 it still has same issue despite having listing for fixed audio streaming
> 
> the mario eclipse dev team said only with the new version of eclipse will it work with nintendont, this was a prep for that upcoming version, dont know if they released the new version already or not.

i guess we wait for version after 1.0.6

**Comment by Drflash55** [comment]
> dont know if they released the new version already or not.

It is still being worked on as we speak, but it is nearing completion.
